pkg/kubelet/kubelet_node_status.go

```diff
@@ -370,13 +369,18 @@ func (kl *Kubelet) tryUpdateNodeStatus(tryNumber int) error {
 	// apiserver cache (the data might be slightly delayed but it doesn't
 	// seem to cause more conflict - the delays are pretty small).
 	// If it result in a conflict, all retries are served directly from etcd.
-	opts := metav1.GetOptions{}
+	var node *v1.Node
+	var err error
 	if tryNumber == 0 {
-		util.FromApiserverCache(&opts)
-	}
-	node, err := kl.heartbeatClient.Nodes().Get(string(kl.nodeName), opts)
-	if err != nil {
-		return fmt.Errorf("error getting node %q: %v", kl.nodeName, err)
+		node, err = kl.nodeInfo.GetNodeInfo(string(kl.nodeName))
```

Comment:
How does GetNodeInfo maintain its current view of the node object?

This really doesn't seem safe. We're computing a patch to send to the server to modify server state based on a local view of the current state. This seems likely to result in patches that are accepted/applied by the server but do not end up with the desired node object persisted. 

As an example, say the apiserver has a condition on the node that doesn't match the kubelet's view of the condition. The kubelet will not include the condition change in the outgoing patch, since from its perspective, nothing has changed in the node.

Reply:
Technically, even without this change, we were getting the node from etcd directly, but from apiserver cache (that is propagated via watch and can be outdated).

IIRC, the nodeInfo is also propagated via watch. So yes - this may be even more delayed. The question is how worse that situation is.
In general, the scenario you described might have been true even before - it's just a bit more probably after that change. Both before and after this change, this is eventually consistent.

Personally, I don't have strong preference either way. I would say this change is fine, but I also see your concerns.

Reply:
a stale/stuck watch from in the kubelet could result in the situation I described indefinitely. 

can we tell if the watch-cache-fed version has at least observed the last successful heartbeat? That would indicate it is relatively up to date

Reply:
Even observed the last successful heartbeat(or even directly get from etcd), it can not prevent `This seems likely to result in patches that are accepted/applied by the server but do not end up with the desired node object persisted.`  Because Get and Patch are not atomic. 

But as @wojtek-t said `Both before and after this change, this is eventually consistent.`

Reply:
Yeah, that can be protected by `compare resource version`

Reply:
> that can be protected by compare resource version

no comparisons that treat it as a non-opaque value. Checking the heartbeat timestamp to see if it matches the last one we set would be ok (would let us know we observed our last update)

Reply:
resonable

Reply:
Done